### PR TITLE
New UI global search

### LINKED ISSUE #39162: 🔎 Advanced search
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**
The current unified search dialog in the top right is quite small and compact. 

**Describe the solution you'd like**
For the powerful things it supports, it would make more sense to have a bit more ways of filtering. E.g. having a left sidebar with filtering which app to focus on, or some detailed selections on owner of file, or date range, or size.

Edit by @sorbaugh (Last check: 04-09-2023): Since the requirements have shifted somewhat over the course of the discussion (and will likely continue to do so), here is a breakdown of the current state in tasks to minimize confusion. I will make an effort to keep this up to date:

Mockup (original comment -> https://github.com/nextcloud/server/issues/39162#issuecomment-1651037351) 
Bildschirmfoto 2023-09-04 um 13 50 04

Frontend Task breakdown:

- [ ] #40238
- [ ] #40239
- [ ] #40240
- [ ] #40237

Backend Task breakdown:
- [ ] #40246
- [ ] #40249
- [ ] #40250

---

**Date**
For things like date range, we have to make sure it works across different apps like Files, Calendar, Contacts, Talk, Mail etc with different types of dates, so we need to define which date it is for the relevant type:
- Files: modification date (probably more relevant than creation date)
- Calendar: event date (probably more relevant than creation date) https://github.com/nextcloud/server/issues/41345
- Deck card: due date (probably more relevant than creation date, if it exists)
- Talk: message date
- Mail: message date https://github.com/nextcloud/mail/issues/8864
- Contacts: ? birthday
- General fallback would be "Modified date"
- etc. …

**People**
The person dropdown could pick from what the "People menu" in the top right also lists, which is both Nextcloud accounts and Contacts. This would then e.g. for Mail also cover the "Recently contacted" address book which is nice.

- Files: owners and sharees
- Calendar: participants https://github.com/nextcloud/server/issues/41345
- Deck card: assignee
- Talk: message poster and/or person replied to
- Mail: sender and recipients https://github.com/nextcloud/mail/issues/8864
- Contacts: the contact itself, possibly manager also
- General fallback would be sharing

We need mockups for this @nextcloud/designers :)

### PR DESCRIPTION
**Pull request by nfebe**
Resolves: #39162

### CODE REVIEW

**Review by skjnldsv — COMMENTED** [review]
Quick code quality review, (while I know this is a very early state :wink: )
- `console.xxxxx` should be avoided and a logger used instead
- Tabs not spaces
- Please format your classes using the BEM format
- Please also document the various sections in your templates [with](https://github.com/nextcloud/server/blob/16975ae45720945776155f026835cfdaf8491383/apps/files/src/components/FileEntry.vue#L36) [html](https://github.com/nextcloud/server/blob/16975ae45720945776155f026835cfdaf8491383/apps/files/src/components/FileEntry.vue#L39) [comments](https://github.com/nextcloud/server/blob/16975ae45720945776155f026835cfdaf8491383/apps/files/src/components/FileEntry.vue#L46)

I won't do a review of the code itself as I assume this will heavily change in the future :)

File: core/src/components/GlobalSearch/SearchFilterChip.vue [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	name: 'SearchFilterChip',
+	components: {
+		CloseIcon,
+	},
+	props: {
+		filter: Object,
+	},
+	methods: {
+		removeChip() {
+			this.$emit('remove-filter', this.filter)
```
Comment (skjnldsv) [inline core/src/components/GlobalSearch/SearchFilterChip.vue:27]: It's easier to use standard event names like `delete`. You don't really need to specify that this is about filters as we'll see the event on the `<FilterChip />` component in template :)

File: core/src/views/GlobalSearchModal.vue [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+						newResults.push({ id: provider.id, provider: provider.name, results: response.data.ocs.data.entries })
+						console.debug('New results', newResults)
+						console.debug('Global search results:', newResults)
+						this.updateResults(newResults)
+					})
+				})
+			} else {
+				this.results = []
+			}
+		},
```
Comment (Fenn-CS) [inline core/src/views/GlobalSearchModal.vue:218]: @artonge Please see if there are any performance concerns here... The method is debounced before call.

File: core/src/views/GlobalSearch.vue [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+<template>
+	<div>
+		<NcButton aria-label="Global search" @click="toggleGlobalSearch">
```
Comment (ChristophWurst) [inline core/src/views/GlobalSearch.vue:24]: l10n

**Comment by Fenn-CS** [comment]
So I spoke with @AndyScherzinger and we going to merge this, but keep the old search as the default. This would allow for the team to be able to see the new ui on tech preview by setting `unified-search.enabled => false` in the config.

Next week, we would have to send in updates to fix these by priority

- [ ] Icon inversion which does not work well now since icons come in white or black instead of one consistent color.
- [ ] Use the new people filter that @marcoambrosini is working (Current people filter not completed)
- [ ] Update back-end API (if ability not already present) to show "Search in App" dynamically.

**Comment by ChristophWurst** [comment]
@Fenn-CS could you please tell us how to test this? The UI looks identical to `master` for me

image

**Comment by Fenn-CS** [comment]
@ChristophWurst set `unified-search.enabled => false` in the config file

**Review by AndyScherzinger — APPROVED** [review]

**Comment by ChristophWurst** [comment]
> @ChristophWurst set `unified-search.enabled => false` in the config file

No change and the config key doesn't exist in the code. Do I need any external apps for this feature?

**Comment by ChristophWurst** [comment]
State is ok for a very bumpy first version hidden behind a feature flag :+1: 

The biggest issue is white icons and text on white background. My theme is set to #C935BA, so buttons are white. The situation might be better with the default color.

- [ ] BUG: Clicking outside the search modal needs two clicks on the magnifier icon to get the modal back.
- [ ] BUG: Some app icons are white on white.
- [ ] BUG: Modified timestamp bubbles change size on hover, have no pointing cursor.
- [ ] BUG: Generally no feedback for *no results*.
- [ ] BUG: *Search people* has white text on white background.
- [ ] BUG: *Search people* has a tiny user bubble, barely wider than the avatar.
- [ ] BUG: *Search people* *No result* goes to the top, overlaps other input.

**Review by ChristophWurst — APPROVED** [review]

**Comment by Fenn-CS** [comment]
Cypress tests keep timing out but are unrelated. Check a couple of tickets with same issues...

**Comment by AndyScherzinger** [comment]
Not up to date with main branch. I guess it needs another rebase and compile 🤔

**Comment by blizzz** [comment]
> Not up to date with main branch. I guess it needs another rebase and compile 🤔

No, there are no conflicts on the compiled assets

**Comment by blizzz** [comment]
CI complains however:

```
ERROR: There is 1 use of triple dot instead of ellipsis:
At line 17 in file /drone/src/build/../core/src/views/GlobalSearchModal.vue
```

Does it refer to https://github.com/nextcloud/server/pull/40823/files#diff-72e59d550e6509240d3c7bad01689aa5d0f9ce1c3b530fb7a45100fedebf37ffR18 ? Should be replaced with … .

**Comment by marcoambrosini** [comment]
@Fenn-CS I pushed the searchable list component for people. To add it to global search modal copy-paste this:

```
<!-- People search popover -->
    <SearchableList :label-text="t('core', 'Search people')"
	    :search-list="peopleSeclectProps.options"
	    :empty-content-text="t('core', 'Not found')">
	    <template #trigger>
		    <NcButton>
			    {{ t('core', 'People') }}
		    </NcButton>
	    </template>
    </SearchableList>
```

With a few tweaks, we should be able to easily adapt this for the apps too

**Comment by Fenn-CS** [comment]
Bringing this down here,  so it remains visible and easily transferable to a new ticket.

TODO

- [ ] Icon inversion which does not work well now since icons come in white or black instead of one consistent color. (Black icons turn white and white icons turn black, leading to to white on white background issues)
- [ ] Use the new people filter (selector) that @marcoambrosini is working on (Current people filter not completed) - DONE, however, `NcPopover` which is used has issues with closing triggers. 
- [ ] Update back-end API (if ability not already present) to show "Search in App" dynamically.

**Comment by ChristophWurst** [comment]
And all of https://github.com/nextcloud/server/pull/40823#issuecomment-1804326745

**Comment by Fenn-CS** [comment]
Thanks @ChristophWurst updated.

**Pull request merged**